src/clj/rems/db/workflow.clj

```diff
@@ -1,67 +1,40 @@
 (ns rems.db.workflow
-  (:require [clj-time.core :as time-core]
-            [clojure.test :refer [deftest is]]
-            [rems.db.catalogue :as catalogue]
-            [rems.db.core :as db]
+  (:require [rems.db.core :as db]
+            [rems.db.licenses :as licenses]
             [rems.db.users :as users]
             [rems.json :as json]
-            [rems.util :refer [update-present]])
-  (:import [org.joda.time DateTime DateTimeZone]))
-
-(defn- parse-db-time [s]
-  (when s
-    (-> (DateTime/parse s)
-        (.withZone DateTimeZone/UTC))))
-
-(deftest test-parse-db-time
-  (is (= nil (parse-db-time nil)))
-  (is (= (time-core/date-time 2019 1 30 7 56 38 627) (parse-db-time "2019-01-30T09:56:38.627616+02:00")))
-  (is (= (time-core/date-time 2015 2 13 12 47 26) (parse-db-time "2015-02-13T14:47:26+02:00"))))
-
-;; TODO: This should probably be in rems.db.licenses.
-(defn- format-license [license]
-  (-> license
-      (select-keys [:type :textcontent :localizations])
-      (assoc :start (parse-db-time (:start license)))
-      (assoc :end (parse-db-time (:end license)))))
-
-(defn- format-workflow
-  [{:keys [id organization owneruserid modifieruserid title workflow start end expired enabled archived licenses]}]
-  {:id id
-   :organization organization
-   :owneruserid owneruserid
-   :modifieruserid modifieruserid
-   :title title
-   :workflow workflow
-   :start start
-   :end end
-   :expired expired
-   :enabled enabled
-   :archived archived
-   :licenses (mapv format-license licenses)})
+            [rems.util :refer [update-present]]))
+
+;; XXX: Overwriting :start and :end from license table with :start and :end
+;;      from workflow_license table seems error-prone - they could at least
+;;      be named differently to avoid confusion.
+;;
+;;      See a related comment in rems.db.licenses regarding the use of
+;;      various start and end times.
+(defn- join-workflow-license-with-license [workflow-license]
+  (-> (:licid workflow-license)
+      licenses/get-license
+      (assoc :start (:start workflow-license)
+             :end (:end workflow-license))))
+
+(defn- get-workflow-licenses [id]
+  (->> {:wfid id}
+       db/get-workflow-licenses
+       (mapv join-workflow-license-with-license)))
 
 (defn- enrich-and-format-workflow [wf]
   (-> wf
-      (update-present :workflow update :handlers #(mapv users/get-user %))
-      format-workflow))
-
-(defn- parse-workflow-body [json]
-  (json/parse-string json))
-
-(defn- parse-licenses [json]
-  (json/parse-string json))
+      (update :workflow #(json/parse-string %))
+      (assoc :licenses (get-workflow-licenses (:id wf)))
+      db/assoc-expired
+      (update-present :workflow update :handlers #(mapv users/get-user %))))
```

Comment:
Since the old workflows have been removed, `:workflow` is always present. This could be changed to `(update-in [:workflow :handlers] #(mapv users/get-user %))`. Though some day there might be a workflow without `:handlers`, in which case `(update :workflow update-present :handlers #(mapv users/get-user %))` is more future proof.

Reply:
Done.